google/cloud/storage/internal/curl_client.cc

```diff
@@ -137,6 +138,45 @@ Status CurlClient::SetupBuilder(CurlRequestBuilder& builder,
   return Status();
 }
 
+template <typename RequestType>
+std::pair<Status, std::unique_ptr<ResumableUploadSession>>
+CurlClient::CreateResumableSessionGeneric(RequestType const& request) {
+  CurlRequestBuilder builder(
+      upload_endpoint_ + "/b/" + request.bucket_name() + "/o", upload_factory_);
+  auto status = SetupBuilder(builder, request, "POST");
+  if (not status.ok()) {
+    return std::make_pair(status, std::unique_ptr<ResumableUploadSession>());
+  }
+  builder.AddQueryParameter("uploadType", "resumable");
+  builder.AddQueryParameter("name", request.object_name());
+  builder.AddHeader("Content-Type: application/json; charset=UTF-8");
+  std::string request_payload;
+  if (request.template HasOption<WithObjectMetadata>()) {
+    request_payload = request.template GetOption<WithObjectMetadata>()
+                          .value()
+                          .JsonPayloadForUpdate();
+  }
+  builder.AddHeader("Content-Length: " +
+                    std::to_string(request_payload.size()));
+  auto payload = builder.BuildRequest().MakeRequest(request_payload);
```

Comment:
Nit: I hadn't looked at this until I was reviewing your other comment, but this isn't a payload; it's an entire HttpResponse (this is further hidden via the use of `auto`). Please rename (without looking at the code, I'm guessing this should probably be done in several spots, but that should be done in its own PR...).

Reply:
Fixed.  Filed #1583 for the others.